### PR TITLE
feat: enable navigation between submenus via arrow keys

### LINKED ISSUE #52055: [Windows] Left/right arrow keys do not navigate between open submenus
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/main/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://www.github.com/electron/electron/issues) for a bug report that matches the one I want to file, without success.

### Electron Version

42.4.1

### What operating system(s) are you using?

Windows

### Operating System Version

Windows 11 25H2

### What arch are you using?

Other (specify below)

### Last Known Working Electron version

n/a

### Does the issue also appear in Chromium / Google Chrome?

I don't know how to test

### Expected Behavior

(Using windows 11 in emulated environment (parallels) on mac arm64 arch)

Steps to reproduce:

Launch an Electron app with a native menu
Press Alt to focus the menu bar — the first menu header is highlighted
Press ↓ to open the File submenu
Press →

Expected: The File submenu closes, the Edit submenu opens, and its first item is selected.

See paint and notepad for examples of windows apps that behave this way.

### Actual Behavior

Actual: Nothing happens. The → key has no effect while a submenu is open.

### Testcase Gist URL

https://gist.github.com/marisademeglio/f11e87e1432956040f94e9070496dfa4

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by mlaurencin**
#### Description of Change

Closes https://github.com/electron/electron/issues/52055

Navigation can now occur between menu bar headers using the left and right arrows keys. Windows conventions were followed when determining situations when such navigation should be disallowed (e.g. navigating to the left when the currently selected item is a submenu).

#### Checklist

- [x] I have built and tested this change
- [x] I have filled out the PR description
- [x] [I have reviewed and verified the changes](https://github.com/electron/governance/blob/main/policy/ai.md)
- [ ] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

Notes: Added functionality to allow navigation between menu bar headers via left and right arrow keys

### CODE REVIEW

File: patches/chromium/feat_enable_sibling_menu_switching_via_keyboard.patch [hunk 1]
```diff
@@ -0,0 +1,144 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Michaela Laurencin <mlaurencin@electronjs.org>
+Date: Thu, 2 Jul 2026 20:32:59 -0400
+Subject: feat: enable sibling menu switching via keyboard
+
+Allow Left/Right arrow keys to move between adjacent top-level menus (e.g. an application menu bar)
+while a menu is open, matching native platform behavior on Windows and Linux.
+Adds MenuDelegate::GetSiblingMenuByDirection, invoked from MenuController when the user presses
+Left/Right at the top level with nothing to open or close in that direction.
```
Comment (nikwen) [inline patches/chromium/feat_enable_sibling_menu_switching_via_keyboard.patch:9]: Can this patch be upstreamed?

Comment (mlaurencin) [reply patches/chromium/feat_enable_sibling_menu_switching_via_keyboard.patch:9]: I think it is possible, but from what I test locally, I don't think Chromium itself would benefit from it. At least I was not able to find a similar menu structure in Chrome on Windows where this functionality was obviously missing. I might be wrong though, so if someone else can come up with an example, that would be a great help.

**Review by jkleinsc — COMMENTED** [review]
API LGTM

File: shell/browser/ui/views/menu_delegate.cc [hunk 2]
```diff
@@ -141,4 +141,24 @@ views::MenuItemView* MenuDelegate::GetSiblingMenu(
   return nullptr;
 }
 
+bool MenuDelegate::GetSiblingMenuByDirection(bool next) {
+  views::MenuButton* button = nullptr;
+  ElectronMenuModel* model = nullptr;
```
Comment (jkleinsc) [inline shell/browser/ui/views/menu_delegate.cc]: This is never used

File: shell/browser/ui/views/menu_bar.cc [hunk 3]
```diff
@@ -107,6 +107,39 @@ bool MenuBar::GetMenuButtonFromScreenPoint(const gfx::Point& screenPoint,
   return false;
 }
 
+bool MenuBar::GetSiblingMenuButtonByDirection(int current_id,
+                                              bool next,
+                                              ElectronMenuModel** menu_model,
```
Comment (jkleinsc) [inline shell/browser/ui/views/menu_bar.cc]: The lone caller to this method doesn't do anything with the modified ElectronMenuModel, so is this parameter really necessary?

File: shell/browser/ui/views/menu_delegate.cc [hunk 4]
```diff
@@ -141,4 +141,24 @@ views::MenuItemView* MenuDelegate::GetSiblingMenu(
...
+  if (!menu_bar_->GetSiblingMenuButtonByDirection(id_, next, &model, &button) ||
+      button->GetID() == id_) {
```
Comment (jkleinsc) [inline shell/browser/ui/views/menu_delegate.cc]: menu_bar_->GetSiblingMenuButtonByDirection already does this check

File: shell/browser/ui/views/menu_delegate.cc [hunk 5]
```diff
@@ -141,4 +141,24 @@ views::MenuItemView* MenuDelegate::GetSiblingMenu(
...
+
+  bool switch_in_progress = !!button_to_open_;
+  // Always update target to open.
+  button_to_open_ = button;
+  // Switching menu asynchronously to avoid crash.
+  if (!switch_in_progress) {
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE, base::BindOnce(&views::MenuRunner::Cancel,
+                                  base::Unretained(menu_runner_.get())));
+  }
```
Comment (jkleinsc) [inline shell/browser/ui/views/menu_delegate.cc:160]: This is duplicated logic that is also in  MenuDelegate::GetSiblingMenu.  It should be extracted to a helper

File: spec/api-menu-keyboard-spec.ts [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+      autoHideMenuBar: false
+    });
+    w.setMenu(buildMenu());
+    await w.loadURL('about:blank');
+    w.show();
+    w.moveTop();
+    w.focus();
+    // Give the compositor/window manager a moment to make the window active so
+    // that robotjs key events are routed to it.
+    await setTimeout(500);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: using setTimeout(s) in tests is a code smell for flaky tests because depending on CI sometimes the timeout isn't enough. If possible we should use the `waitUntil` helper to wait until a certain condition is satisfied.

File: spec/api-menu-keyboard-spec.ts [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    // Click the window's content area to force real foreground activation before we
+    // start sending keys.
+    const bounds = w.getContentBounds();
+    robot.moveMouse(Math.round(bounds.x + bounds.width / 2), Math.round(bounds.y + bounds.height / 2));
+    robot.mouseClick();
+    await setTimeout(300);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+  // Presses a sequence of keys with a delay between each so the native menu can
+  // keep up. Each entry is either a key name or a [key, modifiers] tuple.
+  const pressKeys = async (keys: Array<string | [string, string[]]>) => {
+    for (const key of keys) {
+      if (Array.isArray(key)) {
+        robot.keyTap(key[0], key[1]);
+      } else {
+        robot.keyTap(key);
+      }
+      await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts:101]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+  beforeEach(() => {
+    lastClicked = null;
+  });
+
+  afterEach(async () => {
+    // Ensure any open menu is dismissed before the window is destroyed
+    if (robot && robot.keyTap) {
+      robot.keyTap('escape');
+      robot.keyTap('escape');
+      await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts:138]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+  it('switches to the next sibling menu when Right is pressed on a top-level leaf item', async () => {
+    await createWindow();
+
+    await pressKeys([
+      ['a', ['alt']], // open Alpha (Alpha Item selected)
+      'right', // switch to Beta (next sibling)
+      'enter' // activate Beta's first item
+    ]);
+
+    await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+  it('wraps to the last sibling menu when Left is pressed on the first menu', async () => {
+    await createWindow();
+
+    await pressKeys([
+      ['a', ['alt']], // open Alpha (Alpha Item selected)
+      'left', // switch to Gamma (previous sibling, wrapping)
+      'enter' // activate Gamma's first item
+    ]);
+
+    await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+  it('wraps to the first sibling menu when Right is pressed on the last menu', async () => {
+    await createWindow();
+
+    await pressKeys([
+      ['g', ['alt']], // open Gamma (Gamma Item selected)
+      'right', // switch to Alpha (next sibling, wrapping)
+      'enter' // activate Alpha's first item
+    ]);
+
+    await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    await createWindow();
+
+    await pressKeys([
+      ['b', ['alt']], // open Beta (Beta Leaf selected)
+      'down', // select Beta More (has a submenu)
+      'right', // open the Beta More submenu (NOT switch to Gamma)
+      'enter' // activate Beta More Item
+    ]);
+
+    await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+    await pressKeys([
+      ['b', ['alt']], // open Beta (Beta Leaf selected)
+      'down', // select Beta More (a submenu item)
+      'left', // must NOT switch to Alpha
+      'up', // back to Beta Leaf
+      'enter' // activate Beta Leaf
+    ]);
+
+    await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+
+    await pressKeys([
+      ['b', ['alt']], // open Beta (Beta Leaf selected)
+      'down', // select Beta More (a submenu item)
+      'right', // open the Beta More submenu (Beta More Item selected)
+      'right', // must NOT switch to Gamma (still nested)
+      'enter' // activate Beta More Item
+    ]);
+
+    await setTimeout(KEY_DELAY);
```
Comment (jkleinsc) [inline spec/api-menu-keyboard-spec.ts]: same comment as above about setTimeout vs waitUntil.

File: spec/api-menu-keyboard-spec.ts (hunk 6 above)
Comment (jkleinsc) [reply spec/api-menu-keyboard-spec.ts]: eg 
```
diff --git a/spec/api-menu-keyboard-spec.ts b/spec/api-menu-keyboard-spec.ts
index 7fde930474..b727f4c0e9 100644
--- a/spec/api-menu-keyboard-spec.ts
+++ b/spec/api-menu-keyboard-spec.ts
@@ -5,7 +5,7 @@ import { expect } from 'chai';
 import { setTimeout } from 'node:timers/promises';
 
 import { hasCapturableScreen } from './lib/screen-helpers';
-import { ifdescribe } from './lib/spec-helpers';
+import { ifdescribe, waitUntil } from './lib/spec-helpers';
 import { closeAllWindows } from './lib/window-helpers';
 
 // These tests are skipped when robotjs or a capturable screen is unavailable
@@ -147,7 +147,7 @@ ifdescribe(process.platform !== 'darwin')('menu bar keyboard sibling switching',
       'enter' // activate Beta's first item
     ]);
 
-    await setTimeout(KEY_DELAY);
+    await waitUntil(() => lastClicked !== null);
     expect(lastClicked).to.equal('beta-leaf');
   });
 
@@ -160,7 +160,7 @@ ifdescribe(process.platform !== 'darwin')('menu bar keyboard sibling switching',
       'enter' // activate Gamma's first item
     ]);
 
-    await setTimeout(KEY_DELAY);
+    await waitUntil(() => lastClicked !== null);
     expect(lastClicked).to.equal('gamma-item');
   });
 
@@ -173,7 +173,7 @@ ifdescribe(process.platform !== 'darwin')('menu bar keyboard sibling switching',
       'enter' // activate Alpha's first item
     ]);
 
-    await setTimeout(KEY_DELAY);
+    await waitUntil(() => lastClicked !== null);
     expect(lastClicked).to.equal('alpha-item');
   });
 
@@ -187,7 +187,7 @@ ifdescribe(process.platform !== 'darwin')('menu bar keyboard sibling switching',
       'enter' // activate Beta More Item
     ]);
 
-    await setTimeout(KEY_DELAY);
+    await waitUntil(() => lastClicked !== null);
     expect(lastClicked).to.equal('beta-more-item');
   });
 
@@ -202,7 +202,7 @@ ifdescribe(process.platform !== 'darwin')('menu bar keyboard sibling switching',
       'enter' // activate Beta Leaf
     ]);
 
-    await setTimeout(KEY_DELAY);
+    await waitUntil(() => lastClicked !== null);
     expect(lastClicked).to.equal('beta-leaf');
   });
 
@@ -217,7 +217,7 @@ ifdescribe(process.platform !== 'darwin')('menu bar keyboard sibling switching',
       'enter' // activate Beta More Item
     ]);
 
-    await setTimeout(KEY_DELAY);
+    await waitUntil(() => lastClicked !== null);
     expect(lastClicked).to.equal('beta-more-item');
   });
 });
```

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Overall looks good, just some minor nits.

Also, I do think that this patch probably wouldn't be accepted upstream since Chromium doesn't use top level menus on Windows.

**Review by erickzhao — APPROVED** [review]
API LGTM

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Test should be updated to use `waitUntil` instead of `setTimeout`.  My comment on this didn't render properly, so I have updated it to more clearly explain the change.

**Review by ckerr — COMMENTED** [review]
API LGTM

File: spec/api-menu-keyboard-spec.ts [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+        robot.keyTap(key[0], key[1]);
+      } else {
+        robot.keyTap(key);
+      }
+      await setTimeout(KEY_DELAY);
+    }
+  };
+
+  before(async function () {
+    if (!robot || !robot.keyTap || !hasCapturableScreen()) {
```
Comment (ckerr) [inline spec/api-menu-keyboard-spec.ts]: These tests are intended to cover Linux, but I think they never run on Linux because  `hasCapturableScreen()` is hardcoded to return `false` on Linux CI. We don't need screen capture for this test, and xvfb should be able to handle RobotJS, so let's try enabling the tests on Linux:


```suggestion
    if (!robot?.keyTap) {
```

**Review by ckerr — CHANGES_REQUESTED** [review]

File: spec/api-menu-keyboard-spec.ts (hunk 16 above)
Comment (mlaurencin) [reply spec/api-menu-keyboard-spec.ts]: My understanding though is that the virtual display we have for Linux CI is 0x0, so the focus + mouseClick combination used to ensure each test has proper initial setup would not function. I may be missing an alternative, but I think tests for this functionality cannot be run with our current CI and can only be tested locally.

**Comment by mlaurencin** [comment]
> Test should be updated to use `waitUntil` instead of `setTimeout`. My comment on this didn't render properly, so I have updated it to more clearly explain the change.

I have an updated commit to address the `setTimeout` usage in window creation that I could push, but I think my previous Linux comment pertains to this as well since Windows virtual displays on CI are also 0x0.

**Review by jkleinsc — CHANGES_REQUESTED** [review]
Needs rebase, lint fixed, and patch updated

File: spec/api-menu-keyboard-spec.ts (hunk 16 above)
Comment (jkleinsc) [reply spec/api-menu-keyboard-spec.ts]: @mlaurencin can we try dropping the hasCapturableScreen to verify if that is needed for robotjs to run?

**Review by jkleinsc — CHANGES_REQUESTED** [review]
I think we should at least verify if robotjs can run in CI.

Beyond that it appears from AI review that there are a couple of concerns if this PR is intended to also target Linux and/or if we are going to follow Windows conventions:

### Blocking: sibling switching doesn't work on Linux (or for mouse-opened menus) until an item is manually selected

`PlatformStyle::kAutoSelectFirstMenuItemFromKeyboard` is `BUILDFLAG(IS_WIN)`, so on Linux a keyboard-opened menu leaves the root `MenuItemView` selected. With the root selected:

- **Right** hits the `HasSubmenu()` branch and just selects the first item instead of switching headers.
- **Left** falls through to a no-op `CloseSubmenu()`.

Arrow switching only kicks in after pressing <kbd>Down</kbd> first. The same dead-end applies to mouse-opened menus on **both** platforms (no event / `kMouse` source → no auto-select), which is exactly where native Win32/GTK menus *do* switch headers.

**Suggested fix:** treat a parentless (root) selection as switch-eligible in both `OpenSubmenuOrSwitchSiblingMenu` and `CloseSubmenuOrSwitchSiblingMenu`. Context menus are unaffected since the default delegate returns `false`. Please also add a test that opens a menu without relying on auto-select (e.g. mouse-click the menu bar button, then <kbd>Right</kbd>).

### Behavior vs. Windows conventions — please confirm intent (or update the PR description)

- <kbd>Left</kbd> on a top-level item that has a (closed) flyout does nothing; native Windows moves to the previous menu (item type only matters for <kbd>Right</kbd>).
- <kbd>Right</kbd> on a leaf inside a nested flyout does nothing; native Windows closes the chain and moves to the next top-level menu.
- Edge case: <kbd>Right</kbd> on a top-level item with a *disabled* submenu neither opens nor switches.

### Tests never run on CI

The suite is `ifdescribe(process.platform !== 'darwin')`, but `hasCapturableScreen()` is true on CI only for darwin — so the suite is skipped everywhere on CI. Since nothing here actually captures the screen (the gate is really "robotjs can inject input"), these look runnable on Linux CI — worth an attempt, or a follow-up issue with the TODO reworded. The fixed 350ms key delays remain a flake risk locally.

### Nits

- The inner loop in `MenuBar::GetSiblingMenuButtonByDirection` can be replaced with `GetViewByID(index)`.
- Consider `SwitchToSiblingMenu(bool next)` — `GetSiblingMenuByDirection` is a "Get" with side effects, and since it's a new virtual in the Chromium patch it's worth naming well.
- `//TODO` is missing a space after `//`.

**Comment by mlaurencin** [comment]
> ### Blocking: sibling switching doesn't work on Linux (or for mouse-opened menus) until an item is manually selected
> `PlatformStyle::kAutoSelectFirstMenuItemFromKeyboard` is `BUILDFLAG(IS_WIN)`, so on Linux a keyboard-opened menu leaves the root `MenuItemView` selected. With the root selected:
> 
> * **Right** hits the `HasSubmenu()` branch and just selects the first item instead of switching headers.
> * **Left** falls through to a no-op `CloseSubmenu()`.
> 
> Arrow switching only kicks in after pressing Down first. The same dead-end applies to mouse-opened menus on **both** platforms (no event / `kMouse` source → no auto-select), which is exactly where native Win32/GTK menus _do_ switch headers.
> 
> **Suggested fix:** treat a parentless (root) selection as switch-eligible in both `OpenSubmenuOrSwitchSiblingMenu` and `CloseSubmenuOrSwitchSiblingMenu`. Context menus are unaffected since the default delegate returns `false`. Please also add a test that opens a menu without relying on auto-select (e.g. mouse-click the menu bar button, then Right).

I developed this on Windows, so would need to verify the statements regarding Linux here locally, but I will report back once I do.

> ### Behavior vs. Windows conventions — please confirm intent (or update the PR description)
> * Left on a top-level item that has a (closed) flyout does nothing; native Windows moves to the previous menu (item type only matters for Right).

This was explicitly done. I have been using Microsoft Paint and Notepad to compare with the native Windows application functionality and the closed flyout does nothing. My example here is the top-level item "Zoom" in the View menu in Notepad.

> * Right on a leaf inside a nested flyout does nothing; native Windows closes the chain and moves to the next top-level menu.

Like mentioned above, in my testing they both do nothing - not the closing the chain like is mentioned here. Zoom can also be used as an example here.

> * Edge case: Right on a top-level item with a _disabled_ submenu neither opens nor switches.

I am not entirely sure how to best test the edge case in terms of how native Windows apps handle this scenario. Given how flyouts are handled though, I would think this would be the native functionality as well.

> ### Tests never run on CI
> The suite is `ifdescribe(process.platform !== 'darwin')`, but `hasCapturableScreen()` is true on CI only for darwin — so the suite is skipped everywhere on CI. Since nothing here actually captures the screen (the gate is really "robotjs can inject input"), these look runnable on Linux CI — worth an attempt, or a follow-up issue with the TODO reworded. The fixed 350ms key delays remain a flake risk locally.

hasCapturableScreen has been removed and it appears the Linux tests are passing. My understanding was that the Linux CI does not have screens such that the mouse movement and clicking would register properly, but if it is acceptable that they currently run and pass, then I am fine with them in their current state.


> ### Nits
> * The inner loop in `MenuBar::GetSiblingMenuButtonByDirection` can be replaced with `GetViewByID(index)`.
> * Consider `SwitchToSiblingMenu(bool next)` — `GetSiblingMenuByDirection` is a "Get" with side effects, and since it's a new virtual in the Chromium patch it's worth naming well.
> * `//TODO` is missing a space after `//`.

I can look into these.

**Comment by MarshallOfSound** [comment]
Heads up: Windows CI builds moved to new runners on Monday and the old ones have now been removed. This branch predates that, so its Windows build jobs still point at the old runners and will sit queued forever if CI runs again. Rebasing on `main` (or "Update branch") picks up the change, nothing else needed.

**Comment by rf-figma** [comment]
@mlaurencin since there are still open questions about Linux support etc., is it reasonable to assume this will land in a later minor version and not 44.0?

**Comment by mlaurencin** [comment]
@rf-figma Yes, not shooting to get this in for 44.0.0

File: patches/chromium/feat_enable_sibling_menu_switching_via_keyboard.patch [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
++  const bool is_top_level_child =
++      selected && !selected->HasSubmenu() && selected->GetParentMenuItem() &&
++      !selected->GetParentMenuItem()->GetParentMenuItem();
++  if (is_top_level_child && selected->GetDelegate() &&
++      selected->GetDelegate()->GetSiblingMenuByDirection(/*next=*/true)) {
++    return;
++  }
++  OpenSubmenuChangeSelectionIfCan();
++}
++
```
Comment (MarshallOfSound) [inline patches/chromium/feat_enable_sibling_menu_switching_via_keyboard.patch:62]: `is_top_level_child` requires a parented item without a submenu. Two states never reach `GetSiblingMenuByDirection`.

1. The root is selected. This is the initial state for a menu opened by mouse click. It is also the initial state for a keyboard-opened menu on Linux, where `kAutoSelectFirstMenuItemFromKeyboard` is off. Right only highlights the first item. Left does nothing until the user presses Down.
2. A top-level item has a disabled submenu. It fails `HasSubmenu() && GetEnabled()`. It also fails `!HasSubmenu()`. Right does nothing.

Native Win32 and GTK move to the adjacent header in both cases. Treat a parentless root selection as switch-eligible. Gate the switch on "cannot open": `!HasSubmenu() || !GetEnabled()` in both helpers. Add a test that opens a menu by mouse click and presses Right. The spec only opens menus with Alt+mnemonic.

File: spec/api-menu-keyboard-spec.ts [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    }
+  });
+
+  after(async () => {
+    await closeAllWindows();
+  });
+
+  it('switches to the next sibling menu when Right is pressed on a top-level leaf item', async () => {
+    await pressKeys([
+      ['a', ['alt']], // open Alpha (Alpha Item selected)
```
Comment (MarshallOfSound) [inline spec/api-menu-keyboard-spec.ts:148]: Nit: the `waitUntil` predicate already contains the expected value. The `expect` on the next line can only run when it passes. The likely regression for this feature is that the wrong sibling activates. Then `lastClicked` is a different non-null string. The test stalls for 10 s and reports only "waitUntil timed out after 10000ms".

Use `await waitUntil(() => lastClicked !== null)` and then the `expect`. That fails at once with `expected 'gamma-item' to equal 'alpha-item'`.
